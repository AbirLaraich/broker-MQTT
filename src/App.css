:root {
  --color: rgb(9, 94, 205);
}

.container {
  display: flex;
  flex-direction: column;
  grid-template-columns: 200px 1fr;
  width: 100%;
  height: 200%;
}

.header {
  width: 100%;
  height: 80px;
  background: linear-gradient(to right, #068fff, #ffffff);
  box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
}

.logo {
  height: 80px;
}

.subHeader {
  background: transparent;
  padding: 20px;
  display: flex;
  flex-direction: row;
  border-right: 1px solid #ccc;
  justify-content: center;
}

.text {
  border-radius: 30px;
  height: 30px;
  margin-top: 25px;
}

.subHeaderTitle {
  padding: 0;
  margin-right: 10px;
}


.button {
  height: 40px;
  margin-top: 20px;
  margin-left: 15px;
  display: inline-block;
  padding: .40rem 1.25rem;
  border-radius: 10rem;
  color: #fff;
  text-transform: uppercase;
  font-size: 1rem;
  letter-spacing: .10rem;
  transition: all .3s;
  position: relative;
  overflow: hidden;
  z-index: 1;
}

.button::after {
  content: '';
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: var(--color);
  border-radius: 10rem;
  z-index: -2;
}

.button::before {
  content: '';
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.15);
  transition: all .3s;
  border-radius: 10rem;
  z-index: -1;
}

.button:hover {
  color: #fff;
}

.button:hover::before {
  width: 100%;
}

.navbar {
  width: 100%;
  height: 50px;
  background-color: transparent;
}

.sidebarList {
  display: flex;
  height: 100%;
  flex-direction: row;
  list-style: none;
  padding: 0;
}

.sidebarItem {
  margin-bottom: 10px;
  transition: background-color 0.3s;
}

.sidebarLink {
  color: #333;
  text-decoration: none;
  display: block;
  padding: 5px 10px;
  border-radius: 3px;
}

.sidebarItem:hover {
  background-color: #ece9e9;
  border-radius: 30px;
  box-shadow: rgba(50, 50, 93, 0.25) 0px 6px 12px -2px, rgba(0, 0, 0, 0.3) 0px 3px 7px -3px;
}

.sidebarItem:hover .sidebarLink {
  color: var(--color);
}


.main-content {
  padding: 20px;
}