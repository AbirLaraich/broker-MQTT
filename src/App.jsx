import React, { Component } from 'react';
import { BrowserRouter as Router, Routes, Route, Link } from 'react-router-dom';
import mqtt from 'mqtt';
import SensorPage from './SensorPage';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sensorDataList: {},
      hoveredSensor: null, // État pour stocker le capteur survolé
      mqttUrl: 'wss://random.pigne.org',
      isConnected: false // État pour indiquer si la connexion MQTT est établie
    };
    this.handleUrlChange = this.handleUrlChange.bind(this);
    this.handleToggleConnection = this.handleToggleConnection.bind(this);
  }

  componentWillUnmount() {
    if (this.client) {
      this.client.end();
    }
  }

  connectToMqtt(url) {
    const client = mqtt.connect(url);

    client.on('connect', () => {
      console.log('Connecté au MQTT');
      client.subscribe('value/#');
      this.setState({ isConnected: true });
    });

    client.on('message', (topic, message) => {
      const parsedMessage = JSON.parse(message.toString());
      this.setState(prevState => {
        const updatedList = { ...prevState.sensorDataList };
        if (!updatedList[parsedMessage.name]) {
          updatedList[parsedMessage.name] = [];
        }
        updatedList[parsedMessage.name].push(parsedMessage);
        return { sensorDataList: updatedList };
      });
    });

    this.client = client;
  }

  disconnectFromMqtt() {
    if (this.client) {
      this.client.end();
      this.setState({ isConnected: false, sensorDataList: {} });
    }
  }

  handleMouseEnter(sensorName) {
    this.setState({ hoveredSensor: sensorName });
  }

  handleToggleConnection() {
    if (this.state.isConnected) {
      this.disconnectFromMqtt();
    } else {
      this.connectToMqtt(this.state.mqttUrl);
    }
  }

  handleMouseLeave() {
    this.setState({ hoveredSensor: null });
  }

  handleUrlChange(event) {
    this.setState({ mqttUrl: event.target.value });
  }

  render() {
    return (
      <Router>
        <div style={{ display: 'flex' }}>
          <div style={styles.sidebar}>
            <h2 style={styles.sidebarTitle}>Sensors</h2>
            <input
              type="text"
              value={this.state.mqttUrl}
              onChange={this.handleUrlChange}
              placeholder="MQTT Broker: "
              style={{ marginBottom: '10px' }}
            />
            <button onClick={this.handleToggleConnection}>
              {this.state.isConnected ? 'Disconnect' : 'Connect'}
            </button>
            <ul style={styles.sidebarList}>
              {Object.keys(this.state.sensorDataList).map(sensorName => (
                <li
                  key={sensorName}
                  style={{
                    ...styles.sidebarItem,
                    backgroundColor: this.state.hoveredSensor === sensorName ? '#ddd' : 'transparent'
                  }}
                  onMouseEnter={() => this.handleMouseEnter(sensorName)}
                  onMouseLeave={() => this.handleMouseLeave()}
                >
                  <Link to={`/sensor/${sensorName}`} style={styles.sidebarLink}>{sensorName}</Link>
                </li>
              ))}
            </ul>
          </div>
          <div style={{ flex: 1, padding: '20px' }}>
            <Routes>
              <Route
                exact
                path="/"
                element={<div></div>} 
              />
              <Route
                path="/sensor/:sensorName"
                element={<SensorPage sensorDataList={this.state.sensorDataList} />}
              />
            </Routes>
          </div>
        </div>
      </Router>
    );
  }
}

const styles = {
  sidebar: {
    width: '200px',
    background: '#f0f0f0',
    padding: '20px',
    borderRight: '1px solid #ccc',
  },
  sidebarTitle: {
    margin: '0 0 20px',
    padding: '0',
  },
  sidebarList: {
    listStyle: 'none',
    padding: 0,
  },
  sidebarItem: {
    marginBottom: '10px',
    transition: 'background-color 0.3s',
  },
  sidebarLink: {
    color: '#333',
    textDecoration: 'none',
    display: 'block',
    padding: '5px 10px',
    borderRadius: '3px',
  },
};

export default App;
