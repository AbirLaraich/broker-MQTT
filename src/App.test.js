import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import App from './App';
import mqtt from 'mqtt';

jest.mock('mqtt', () => {
  const mockMqtt = {
    connect: jest.fn(() => ({
      on: jest.fn((event, callback) => {
        if (event === 'connect') {
          callback();
        }
      }),
      subscribe: jest.fn(),
      end: jest.fn()
    }))
  };
  return mockMqtt;
});

describe('App', () => {
  test('Le bouton de connexion MQTT', () => {
    const { getByText } = render(<App />);
    const connectButton = getByText('Connect');
    expect(connectButton).toBeDefined();
  });

  test('Activer la connexion MQTT en cliquant sur un bouton', () => {
    const { getByText } = render(<App />);
    const connectButton = getByText('Connect');
    fireEvent.click(connectButton);
    expect(mqtt.connect).toHaveBeenCalled();
    expect(connectButton.textContent).toBe('Disconnect');
    fireEvent.click(connectButton);
    expect(connectButton.textContent).toBe('Connect');
  });

  test('Modifier l\'URL MQTT en cas de changement d\'entrée', () => {
    const { getByPlaceholderText } = render(<App />);
    const urlInput = getByPlaceholderText('url');
    fireEvent.change(urlInput, { target: { value: 'wss://example.com' } });
    expect(urlInput.value).toBe('wss://example.com');
  });

  test('Mise à jour l\'état de mqttUrl en cas de modification de l\'entrée de l\'URL', () => {
    const { getByPlaceholderText } = render(<App />);
    const urlInput = getByPlaceholderText('url');
    fireEvent.change(urlInput, { target: { value: 'wss://example.com' } });
    expect(urlInput.value).toBe('wss://example.com');
  });

  test('Afficher les capteurs dans sidebar', () => {
    const sensorDataList = {
      sensor1: [],
      sensor2: [],
    };
    const { container } = render(<App />, { initialState: { sensorDataList } });
    expect(container.firstChild).toMatchSnapshot();
  });
});
