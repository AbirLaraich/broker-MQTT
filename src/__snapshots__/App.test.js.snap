// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Afficher les capteurs dans sidebar 1`] = `
<div
  class="container"
>
  <div
    class="header"
  >
    <img
      class="logo"
      src="../src/assets/sensor.png"
    />
  </div>
  <div
    class="subHeader"
  >
    <h2
      class="subHeaderTitle"
    >
      MQTT Broker:
    </h2>
    <input
      class="text"
      placeholder="url"
      type="text"
      value="wss://random.pigne.org"
    />
    <button
      class="button"
    >
      Connect
    </button>
  </div>
  <div
    class="navbar"
  >
    <ul
      class="sidebarList"
    />
  </div>
  <div
    class="main-content"
  >
    <div />
  </div>
</div>
`;
